### ISSUE TITLE
Investigate using opentracing standard for wsmaster and wsagent logging

### ISSUE THREAD

**Issue by l0rd** [redhat-developer/rh-che#687]
This is for supporting on prem installations, not for osio. We should also think about how does this fit in current osio infrastructure.

http://opentracing.io/documentation/pages/api/

**Comment by l0rd** [comment]
The output of this task should be not a working che with opentracing support but just a comment on the amount of work needed to have it (a PoC would be a nice to have too).

**Comment by ibuziuk** [comment]
@l0rd @slemeur @gorkem I have started to investigate this issue and have a couple of point that I would like to discuss (I believe the output of this rh-che  task would be creating an epic upstream with sub tasks / estimations):

- am I correct that in the end it is planned to have some property e.g. `che.opentracing.tracer` which would allow to specify particular tracer e.g. `jaeger` / `zipkin`. The tracer config would be implemented in che and depending on tracer that is specified the right configuration would be picked up during the runtime.
- user will have the possibility to specify tracer and tracer options only via properties / env vars
- every new tracer support would need to be contributed to che. e.g. we could originally support only jaeger and later add zipkin or any other tracer. The point here is that every new tracer support would require contibuting new configuration / tracer specific lib to che.
- wsmaster & wsagent are expected to use the same configuration but separate service names in
 tracer

WDYT ?

**Comment by ibuziuk** [comment]
Just to be more specific I have created a POC with support of tracer picking in che-starter [1]. Basically, there is property [2] for specifying the tracer (`zipkin` & `jaeger` are supported). The tracer config is picked up during the runtime [3] depending on the tracer that is specified in property / env var. It is possible to test locally by running jaeger & zipkin via:

`docker run --rm -it --network=host jaegertracing/all-in-one
`
`docker run --rm -it -p 9411:9411 openzipkin/zipkin`

Than running che-starter with a particular tracer set - all requests would be collected by the tracer that was selected during che-starter startup. Once che-starter is restarted with another tracer specified all new requests would be collected by another tracer.

Am I correct that smth. similar is expected to be implemented in che ?

[1] https://github.com/ibuziuk/che-starter/tree/opentracing
[2] https://github.com/ibuziuk/che-starter/commit/01d17213c766b10aa1a56970fb4299678e563a18#diff-660a4c1a5cbedaa6e2556dde19faeb2dR23
[3] https://github.com/ibuziuk/che-starter/commit/01d17213c766b10aa1a56970fb4299678e563a18#diff-09bc96bae1d1177874938a1a807f7342R39

**Comment by l0rd** [comment]
I have just seen a demo of @ibuziuk PoC and it looks promising. Really cool stuff. What you should continue investigating is:
- How to enrich opentracing data with application logs (e.g. slf4j)
- How to query for opentracing data coming from distinct services (e.g. che-starter and wsmaster)
- Record a video that shows the data that jaeger collects from wsmaster and che-starter when a workspace is started

**Comment by gorkem** [comment]
> Record a video that shows the data that jaeger collects from wsmaster and che-starter when a workspace is started 

How about data from the workspace that has just started?

**Comment by ibuziuk** [comment]
@gorkem just to simplify the initial POC it was decided to omit wsagent interactions for now

**Comment by garagatyi** [comment]
Eventually, we will replace current wsagent, so we should consider how to use Open Tracing stuff with Theia, Language servers, and any other cloud IDE/tooling. So, I would omit the addition of something specific to current wsagent on Che master.

**Comment by ibuziuk** [comment]
> How to enrich opentracing data with application logs (e.g. slf4j)

So far I was not able to find any opentracing configuration option that would allow to add the application logs to the tracer automatically by default. Basically, it seems that it is expected that dev would add extra logs to the tracer programmatically if needed. Also, after investigating tracing more it appeared to be that normally `logging` and `tracing` (particular case of `instrumentation`) are handled separately and adding all the application logs to the tracer  does not seem to be a good idea (tracer should mainly include sample data and tags for request identification which would be used for querying).

- https://github.com/openzipkin/zipkin/issues/1453
- https://peter.bourgon.org/blog/2016/02/07/logging-v-instrumentation.html

In terms of correlation between logs and traces while having those two decoupled, there is an interesting comment which describes the approach of adding `trace ID` to every log which allows to query all log messages associated with a given `trace ID`   - https://github.com/openzipkin/zipkin/issues/1453#issuecomment-272315347

More details about the investigation of adding application logs to tracer can be found here -  https://stackoverflow.com/questions/50855480/how-to-enrich-jaeger-opentracing-data-with-the-application-logs-produced-by-slf

**Comment by ibuziuk** [comment]
> How to query for opentracing data coming from distinct services (e.g. che-starter and wsmaster)
for querying data span tags shared across services should be used e.g. `req_id` /  `identity_id` / `traceId` etc. 
e.g. searching traces between che-starter & wsmaster based on `request_id`:

![image](https://user-images.githubusercontent.com/1461122/41612676-e79baa44-73f3-11e8-818a-7e1c3c69ee59.png)

![image](https://user-images.githubusercontent.com/1461122/41612724-07957488-73f4-11e8-8a6b-d91f87d65971.png)

> Record a video that shows the data that jaeger collects from wsmaster and che-starter when a workspace is started
https://youtu.be/4tWeH8JqQQk

**Comment by ibuziuk** [comment]
Subtasks with estimations that could potentially be added to the next sprint:
- creating CQs for opentracing & jaeger and resolving dependency conflicts - `5`
- adding tracer agnostic  opentracing support to wsmaster - `21` (delivery of this would most likely be blocked by CQs)
